### PR TITLE
Fastpitch e2e

### PR DESCRIPTION
**Pull request by treacker**
# What does this PR do ?

FastPitch + HiFi-GAN end-2-end model

**Collection**: [TTS]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
+            if len(ids_str_max[ids_str_max < 0]) > 0:
+                print('Encountered short batch')
+                return
+
+            mels_pred, ids_slice = rand_slice_segments(
+                mels_pred, dec_lens, self._cfg.segment_size // self._cfg.n_window_stride
+            )
+            audio = slice_segments(audio, ids_slice * self.cfg.n_window_stride, self._cfg.segment_size)
+            mels = slice_segments(mels, ids_slice, self._cfg.segment_size // self._cfg.n_window_stride)
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:518]: Is this different from the "mels_pred" value that is returned by `rand_slice_segments`?

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+        if self.cfg.use_patches:
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
+            if len(ids_str_max[ids_str_max < 0]) > 0:
+                print('Encountered short batch')
+                return
+
+            mels_pred, ids_slice = rand_slice_segments(
+                mels_pred, dec_lens, self._cfg.segment_size // self._cfg.n_window_stride
+            )
+            audio = slice_segments(audio, ids_slice * self.cfg.n_window_stride, self._cfg.segment_size)
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:517]: Nitpick: it will be easier to read if values like `self._cfg.segment_size // self._cfg.n_window_stride` and `ids_slice * self.cfg.n_window_stride` are saved into variables with appropriate names.

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+        if durs is None:
+            durs = attn_hard_dur
+
+        audio = audio.unsqueeze(1)
+
+        fp_inter_mel = self.mel_loss_fn(spect_predicted=mels_pred, spect_tgt=mels)
+
+        # Skip batch if it is too short
+        if self.cfg.use_patches:
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:509]: What is "ids_str" short for?

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 4]
```diff
@@ -0,0 +1,873 @@
...
+
+        self.manual_backward(loss_d)
+        norm_d = clip_grad_value_(
+            itertools.chain(self.hfg_model.msd.parameters(), self.hfg_model.mpd.parameters()), 20
+        )
+        optim_d.step()
+
+        # Train generator
+        optim_g.zero_grad()
+        loss_mel = self.mel_loss_fn(spect_predicted=audio_pred_mel, spect_tgt=mels)
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:547]: Using MelLoss() here means we are using L2 loss instead of L1 loss, which is not really a safe change to make without re-tuning the loss factor (self.l1_factor) of 45.

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+        )
+        msd_score_real, msd_score_gen, _, _ = self.hfg_model.msd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_msd, _, _ = self.discriminator_loss(
+            disc_real_outputs=msd_score_real, disc_generated_outputs=msd_score_gen
+        )
+        loss_d = loss_disc_msd + loss_disc_mpd
+
+        self.manual_backward(loss_d)
+        norm_d = clip_grad_value_(
+            itertools.chain(self.hfg_model.msd.parameters(), self.hfg_model.mpd.parameters()), 20
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:541]: Can we use the `trainer.gradient_clip_val` value we usually have in our config instead of hardcoding to 20?

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 6]
```diff
@@ -0,0 +1,873 @@
...
+
+    @property
+    def input_types(self):
+        return self._input_types
+
+    @property
+    def output_types(self):
+        return self._output_types
+
+    def input_example(self, max_batch=1, max_dim=44):
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:764]: For the first version I would leave out the input_example(), create_batch(), and interpolate_speakers() functions unless we have an immediate need for them and have tested them.

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 7]
```diff
@@ -0,0 +1,873 @@
...
+
+        self.log_dict(metrics, on_step=True, sync_dist=True)
+
+    def validation_step(self, batch, batch_idx):
+        # return
+        attn_prior, durs, speaker, energy = None, None, None, None
+        if self.learn_alignment:
+            assert self.ds_class_name == "TTSDataset", f"Unknown dataset class: {self.ds_class_name}"
+            batch_dict = process_batch(batch, self._validation_dl.dataset.sup_data_types_set)
+            # print(batch_dict.keys())
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:616]: Stray print

File: docs/source/tts/models.rst [hunk 8]
```diff
@@ -124,6 +124,12 @@ VITS is an end-to-end speech synthesis model, which generates raw waveform audio
         :scale: 25%
 
 
+FastPitch End-To-End
+~~~~~~~~~~~~~~~
+This is an end-to-end joint model, which combines FastPitch spectrogram generator model with HiFi-GAN vocoder. This allows it to train and infer as a single model.
+
+
+
```
Comment (rlangman) [inline docs/source/tts/models.rst:132]: Nitpick: the surrounding of the documentation uses 2 blank lines instead of 3

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 9]
```diff
@@ -0,0 +1,873 @@
...
+    add_blank_at: bool = True
+    g2p: G2PConfig = G2PConfig()
+
+
+@dataclass
+class TextTokenizerConfig:
+    text_tokenizer: TextTokenizer = TextTokenizer()
+
+
+class FastPitchE2EModel(TextToWaveform):
```
Comment (rlangman) [inline nemo/collections/tts/models/fastpitch_e2e.py:93]: Please mark the model `@experimental`

File: examples/tts/conf/fastpitch_e2e.yaml [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+phoneme_dict_path: "scripts/tts_dataset_files/ipa_cmudict-0.7b_nv23.01.txt"
+heteronyms_path: "scripts/tts_dataset_files/heteronyms-052722"
+
+model:
+  learn_alignment: true
+  bin_loss_warmup_epochs: 100
+  use_patches: true
+  inter_mel_loss: false
+  use_pretrain: true
```
[the remark this answers is not in the document]
Comment (treacker) [reply examples/tts/conf/fastpitch_e2e.yaml:40]: No, it can be trained from scratch, but I am not sure if it will work smoothly

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 11]
```diff
@@ -0,0 +1,873 @@
...
+        audio = audio.unsqueeze(1)
+
+        fp_inter_mel = self.mel_loss_fn(spect_predicted=mels_pred, spect_tgt=mels)
+
+        # Skip batch if it is too short
+        if self.cfg.use_patches:
+            ids_str_max = dec_lens - self._cfg.segment_size // self._cfg.n_window_stride + 1
+            if len(ids_str_max[ids_str_max < 0]) > 0:
+                print('Encountered short batch')
+                return
```
[the remark this answers is not in the document]
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:511]: It is one way to resolve it, but it will lay responsibility on users to check if min_duration value is correct, since it can be different for different preprocessing parameters. In case of short batch the code will just crash, so I added this check

File: nemo/collections/tts/models/fastpitch_e2e.py (hunk 5 above)
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:540]: 1. We can't, since the code is written with automatic_optimization=False. because it is a GAN model. 
2. I think this is a redundant clipping, was temporary solution which gave 0 effect

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 12]
```diff
@@ -0,0 +1,873 @@
...
+                ),
+            )
+
+            generator = instantiate(cfg.generator)
+            mpd = MultiPeriodDiscriminator(debug=False)
+            msd = MultiScaleDiscriminator(debug=False)
+
+            self.hfg_model.generator = generator
+            self.hfg_model.mpd = mpd
+            self.hfg_model.msd = msd
```
[the remark this answers is not in the document]
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:225]: Vladimir Bataev told me that this is the only way to make it loadable from trained checkpoint. I am not sure it is exactly true, I didn't have issues with it, but maybe it holds for some cases.

File: nemo/collections/tts/models/fastpitch_e2e.py (hunk 2 above)
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:516]: This code is adopted from VITS implementation, do u think we need to change it in both places then?

File: nemo/collections/tts/models/fastpitch_e2e.py (hunk 1 above)
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:517]: `rand_slice_segments` returns indexes for each sample in batch, so we can take same slices from GT mels and audios

File: examples/tts/conf/fastpitch_e2e.yaml [hunk 13]
```diff
@@ -0,0 +1,261 @@
+# This config contains the default values for training a FastPitch model with aligner.
```
[the remark this answers is not in the document]
Comment (treacker) [reply examples/tts/conf/fastpitch_e2e.yaml:1]: I think aligner is still necessary, since I use native FP durations in my last version

File: nemo/collections/tts/models/fastpitch_e2e.py (hunk 4 above)
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:571]: I don't think scales of losses will be different in this case

File: nemo/collections/tts/models/fastpitch_e2e.py (hunk 5 above)
Comment (rlangman) [reply nemo/collections/tts/models/fastpitch_e2e.py]: I meant that the model class could access the value `self.trainer.gradient_clip_val` directly (or get it elsewhere in the config). But if the clip value is not necessary, then it does not matter.

File: nemo/collections/tts/models/fastpitch_e2e.py [hunk 14]
```diff
@@ -0,0 +1,784 @@
...
+            self.register_nemo_submodule(
+                "hfg_model",
+                config_field="hfg_model",
+                model=HifiGanModel.from_pretrained(cfg.fastpitch_model, map_location=torch.device("cpu")),
+            )
+
+            self.register_nemo_submodule(
+                "fp_model",
+                config_field="fp_model",
+                model=FastPitchModel.from_pretrained(cfg.hifigan_mode, map_location=torch.device("cpu")),
```
[the remark this answers is not in the document]
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:182]: `register_nemo_submodule` is used, because pretrained branch uses NeMo Model classes, but for training from scratch there is just plain torch class

[the remark this answers is not in the document]
Comment (treacker) [reply nemo/collections/tts/models/fastpitch_e2e.py:182]: FastPitch uses L2 loss, so for consistency GAN-loss was chosen the same. I think we can switch to L1 without problems

File: tutorials/tts/FastPitch_E2E_Training.ipynb [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "ZivXzmq0YYLj"
+   },
+   "source": [
+    "# FastPitch End-to-end and NeMo\n",
+    "\n",
+    "FastPitch End-to-end is a joint model which combines FastPitch and HiFi-GAN to converts text characters directly into an audio sample. \n",
```
Comment (redoctopus) [inline tutorials/tts/FastPitch_E2E_Training.ipynb:89]: s/converts/convert

File: tutorials/tts/FastPitch_E2E_Training.ipynb [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "Upv_LxBIsC51"
+   },
+   "source": [
+    "# Runner\n",
+    "\n",
+    "Let's take a look at the fastpitch_e2e.py file\n",
+    "\n",
```
Comment (redoctopus) [inline tutorials/tts/FastPitch_E2E_Training.ipynb:171]: The following sections are very similar to the VITS tutorial, can you deduplicate the explanations? Also need to remove some references to VITS.
